infrastructure/src/main/java/org/corfudb/infrastructure/AutoCommitService.java

```diff
@@ -0,0 +1,217 @@
+package org.corfudb.infrastructure;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.protocols.wireprotocol.Token;
+import org.corfudb.protocols.wireprotocol.TokenResponse;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.exceptions.NetworkException;
+import org.corfudb.runtime.exceptions.TrimmedException;
+import org.corfudb.runtime.view.Address;
+import org.corfudb.runtime.view.Layout;
+import org.corfudb.util.CFUtils;
+import org.corfudb.util.LambdaUtils;
+import org.corfudb.util.Sleep;
+import org.corfudb.util.concurrent.SingletonResource;
+
+import java.time.Duration;
+import java.util.Collections;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * An auto-commit service that periodically commits the unwritten addresses
+ * in the global log, continuously consolidating the log prefix.
+ * <p>
+ * Created by WenbinZhu on 5/5/20.
+ */
+@Slf4j
+public class AutoCommitService implements ManagementService {
+
+    private static final int COMMIT_BATCH_SIZE = 500;
+    private static final int MAX_COMMIT_RETRY = 8;
+    private static final Duration CONN_RETRY_RATE = Duration.ofMillis(500);
+
+    private final ServerContext serverContext;
+    private final SingletonResource<CorfuRuntime> runtimeSingletonResource;
+    private final ScheduledExecutorService autoCommitScheduler;
+
+    // Global log tail fetched at last auto commit cycle, which
+    // would be the commit upper bound in the current cycle.
+    private long lastLogTail = Address.NON_ADDRESS;
+    // Cached committed tail so that we do not need to fetch every cycle.
+    private long committedTail = Address.NON_ADDRESS;
+
+    AutoCommitService(@NonNull ServerContext serverContext,
```

Comment:
Why do we need to run it as a separate service, why we can't execute it in RemoteMonitoringService after the iteration of failure detection?

Reply:
1) RemoteMonitoringService executes too often, auto commit needs its own interval
2) Auto commit can block RemoteMonitoringService
3) It's cleaner to have a separate service